runtime/gc_realtime/RealtimeMarkingScheme.cpp

```diff
@@ -480,502 +83,89 @@ void
-	UDATA *descriptionPtr;
-	UDATA descriptionBits;
-	UDATA descriptionIndex;
-#if defined(J9VM_GC_LEAF_BITS)
-	UDATA *leafPtr;
-	UDATA leafBits;
-#endif /* J9VM_GC_LEAF_BITS */
-	
-#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
-	if(_realtimeGC->getRealtimeDelegate()->isDynamicClassUnloadingEnabled()) {
-		markClassOfObject(env, objectPtr);
-	}
-#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
-	
-	descriptionPtr = (UDATA *)J9GC_J9OBJECT_CLAZZ(objectPtr)->instanceDescription;
-#if defined(J9VM_GC_LEAF_BITS)
-	leafPtr = (UDATA *)J9GC_J9OBJECT_CLAZZ(objectPtr)->instanceLeafDescription;
-#endif /* J9VM_GC_LEAF_BITS */
-
-	if(((UDATA)descriptionPtr) & 1) {
-		descriptionBits = ((UDATA)descriptionPtr) >> 1;
-#if defined(J9VM_GC_LEAF_BITS)
-		leafBits = ((UDATA)leafPtr) >> 1;
-#endif /* J9VM_GC_LEAF_BITS */
-	} else {
-		descriptionBits = *descriptionPtr++;
-#if defined(J9VM_GC_LEAF_BITS)
-		leafBits = *leafPtr++;
-#endif /* J9VM_GC_LEAF_BITS */
-	}
-	descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
+	/* Heap Marking and barrier processing. Cannot delay barrier processing until the end.*/	
+	_realtimeGC->completeMarking(realtimeEnv);
 
-	I_32 referenceState = J9GC_J9VMJAVALANGREFERENCE_STATE(env, objectPtr);
-	UDATA referenceObjectType = J9CLASS_FLAGS(J9GC_J9OBJECT_CLAZZ(objectPtr)) & J9AccClassReferenceMask;
-	UDATA referenceObjectOptions = env->_cycleState->_referenceObjectOptions;
-	bool isReferenceCleared = (GC_ObjectModel::REF_STATE_CLEARED == referenceState) || (GC_ObjectModel::REF_STATE_ENQUEUED == referenceState);
-	bool referentMustBeMarked = isReferenceCleared;
-	bool referentMustBeCleared = false;
+	metronomeDelegate->flushRememberedSet(realtimeEnv);
 
-	switch (referenceObjectType) {
-	case J9AccClassReferenceWeak:
-		referentMustBeCleared = (0 != (referenceObjectOptions & MM_CycleState::references_clear_weak));
-		break;
-	case J9AccClassReferenceSoft:
-		referentMustBeCleared = (0 != (referenceObjectOptions & MM_CycleState::references_clear_soft));
-		referentMustBeMarked = referentMustBeMarked || (
-			((0 == (referenceObjectOptions & MM_CycleState::references_soft_as_weak))
-			&& ((UDATA)J9GC_J9VMJAVALANGSOFTREFERENCE_AGE(env, objectPtr) < _gcExtensions->getDynamicMaxSoftReferenceAge())));
-		break;
-	case J9AccClassReferencePhantom:
-		referentMustBeCleared = (0 != (referenceObjectOptions & MM_CycleState::references_clear_phantom));
-		break;
-	default:
-		Assert_MM_unreachable();
+	if (realtimeEnv->_currentTask->synchronizeGCThreadsAndReleaseMaster(realtimeEnv, UNIQUE_ID)) {
+		metronomeDelegate->setUnmarkedImpliesCleared();
+		realtimeEnv->_currentTask->releaseSynchronizedGCThreads(realtimeEnv);
 	}
 
-	GC_SlotObject referentPtr(_javaVM->omrVM, &J9GC_J9VMJAVALANGREFERENCE_REFERENT(env, objectPtr));
+	MM_EnvironmentRealtime *realtimeEnv = MM_EnvironmentRealtime::getEnvironment(env);
+	/* Process reference objects and finalizable objects. */
+	_realtimeGC->getRealtimeDelegate()->markLiveObjectsScan(realtimeEnv);
```

Comment:
This should be called earlier. Before the sync point. And the only thing it should call is flushReferenceObjects.

Reply:
Again, this would change the original order of the code and break it.

Reply:
it will not the break the order if you change what it does, it should be exactly at the spot what now flushReferenceObjects is called

Reply:
as I explained curently markLiveObjectsScan in delegate does the wrong thing. This is the main scan work not mean for any of root passes